File: 23.merge-k-sorted-lists.java
import java.util.Comparator;
import java.util.PriorityQueue;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
	public ListNode mergeKLists(ListNode[] lists) {
		if (lists.length == 0 || lists == null) {
			return null;
		}
		PriorityQueue<ListNode> pq = new PriorityQueue<>(lists.length, new Comparator<ListNode>() {
			@Override
			public int compare(ListNode l1, ListNode l2) {
				return l1.val - l2.val;
			}
		});

		ListNode dummy = new ListNode(0);
		ListNode tail = dummy;

		for (ListNode list : lists) {
			if (list != null) {
				pq.add(list);
			}
		}

		while (!pq.isEmpty()) {
			tail.next = pq.poll();
			tail = tail.next;
			if (tail.next != null) {
				pq.add(tail.next);
			}
		}
		return dummy.next;
	}
}

File: 23.merge-k-sorted-lists.js
/*
 * @lc app=leetcode id=23 lang=javascript
 *
 * [23] Merge k Sorted Lists
 */
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */

// 方法一：Brute Froce
var mergeKLists = function (lists) {

};

// 方法二；Compare one by one
var mergeKLists = function (lists) {
  let min_index = 0;
  let head = new ListNode(0);
  let h = head;
  while (true) {
    let isBreak = true;
    let min = Number.MAX_VALUE;
    for (let i = 0; i < lists.length; i++) {
      if (lists[i]) {
        if (lists[i].val < min) {
          min_index = i;
          min = lists[i].val;
        }
        isBreak = false;
      }
    }

    if (isBreak) {
      break;
    }

    h.next = lists[min_index];
    h = h.next;
    lists[min_index] = lists[min_index].next;
  }

  h.next = null;
  return head.next;
};

